### PR TITLE
Removed unnecessary `factories.array` call

### PR DESCRIPTION
**Pull request by Dhruv454000**
## Description

Replacement of factories.array with DNDarray.

Issue/s resolved: #801

## Changes proposed:

- replace factories.array

## Type of change

- enhancement

## Due Diligence
- [ ] All split configurations tested
- [ ] Multiple dtypes tested in relevant functions
- [ ] Documentation updated (if needed)
- [ ] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
 no

### CODE REVIEW

**Comment by mtar** [comment]
GPU cluster tests are currently disabled on this Pull Request.

File: heat/core/sanitation.py [hunk 1]
```diff
@@ -381,6 +381,13 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
     x : DNDarray
         with `x.ndim = 0`
     """
-    return factories.array(
-        x.larray.unsqueeze(0), dtype=x.dtype, split=x.split, comm=x.comm, device=x.device
+    a = (x.larray.unsqueeze(0),)
```
Comment (mtar) [inline heat/core/sanitation.py:384]: This leaves you with a one element tuple, but we need the tensor itself.

**Comment by mtar** [comment]
I converted this PR to a draft as it is not ready yet.

**Comment by ClaudiaComito** [comment]
Hi @Dhruv454000 , thanks for taking this on. 

Make sure you check out the documentation for [`heat.array`](https://heat.readthedocs.io/en/latest/autoapi/heat/core/factories/index.html#heat.core.factories.array) and the [DNDarray](https://heat.readthedocs.io/en/latest/autoapi/heat/core/dndarray/index.html) class. 

What we're trying to do here is identify those cases in which we use `ht.array` internally in the code, but could be calling the DNDarray class directly (because we already know or can calculate all the metadata, especially `gshape`).

In some cases, `ht.array` requires communication and synchronization among all processes (see the `is_split` argument) and by not using it internally unless it's absolutely necessary, we want to improve efficiency.  

I hope this helps!

**Comment by Dhruv454000** [comment]
> Hi @Dhruv454000 , thanks for taking this on.
> 
> Make sure you check out the documentation for [`heat.array`](https://heat.readthedocs.io/en/latest/autoapi/heat/core/factories/index.html#heat.core.factories.array) and the [DNDarray](https://heat.readthedocs.io/en/latest/autoapi/heat/core/dndarray/index.html) class.
> 
> What we're trying to do here is identify those cases in which we use `ht.array` internally in the code, but could be calling the DNDarray class directly (because we already know or can calculate all the metadata, especially `gshape`).
> 
> In some cases, `ht.array` requires communication and synchronization among all processes (see the `is_split` argument) and by not using it internally unless it's absolutely necessary, we want to improve efficiency.
> 
> I hope this helps!

Ok , I will try.

**Comment by Dhruv454000** [comment]
srry for this commit , there was a confusion

**Comment by Dhruv454000** [comment]
@ClaudiaComito I have checked the docs, In this module I have checked other places where fact.array was present , but they had copy=false so I didnt change them , I saw about is_split also , but didnt get any idea, If I am wrong at any part pls correct me.

**Comment by Dhruv454000** [comment]
@ClaudiaComito I have checked the docs, In this module I have checked other places where fact.array was present , but they had copy=false so I didnt change them , I saw about is_split also , but didnt get any idea, If I am wrong at any part pls correct me.

**Comment by mtar** [comment]
> @ClaudiaComito I have checked the docs, In this module I have checked other places where fact.array was present , but they had copy=false so I didnt change them , I saw about is_split also , but didnt get any idea, If I am wrong at any part pls correct me.

Thank you very much @Dhruv454000, @ClaudiaComito wanted to make you cautious when you're able to replace `factories.array`. If you look at the `split` parameter in `ht.array` and `DNDarray`, you'll see that the replacement you did here wouldn't be possible normally.

**Comment by mtar** [comment]
A check beforehand would be good. This way, we can be sure that we're only dealing with scalars.

**Comment by Dhruv454000** [comment]
> A check beforehand would be good. This way, we can be sure that we're only dealing with scalars.

Done.

**Comment by Dhruv454000** [comment]
> A check beforehand would be good. This way, we can be sure that we're only dealing with scalars.

Done

**Comment by Dhruv454000** [comment]
@ClaudiaComito I have added a check to confirm it is a scalar dndarray , Is anything wrong over there?

File: heat/core/sanitation.py [hunk 2]
```diff
@@ -381,6 +381,17 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
...
+        )
+    a = x.larray.unsqueeze(0)
+    return DNDarray(
+        a,
+        gshape=tuple(a.size()),
+        dtype=x.dtype,
+        split=x.split,
+        comm=x.comm,
+        device=x.device,
+        balanced=x.is_balanced,
```
Comment (ClaudiaComito) [inline heat/core/sanitation.py:396]: We can set `gshape = (1,), split=None, balanced=True`, we know it's a 1-element array and scalar DNDarrays are non-distributed by definition.

**Comment by ClaudiaComito** [comment]
> 



> @ClaudiaComito I have added a check to confirm it is a scalar dndarray , Is anything wrong over there?

@Dhruv454000 no nothing wrong, we can simplify things a bit (see above). When you're done with that, we can run the GPU tests and check if the test coverage is sufficient. 

Also could you please update the changelog? There's a similar entry already, you can add to that one.

File: heat/core/sanitation.py (hunk 2 above)
Comment (Dhruv454000) [reply heat/core/sanitation.py:396]: > We can set `gshape = (1,), split=None, balanced=True`, we know it's a 1-element array and scalar DNDarrays are non-distributed by definition.

Yes , got it.

**Comment by Dhruv454000** [comment]
> > 
> 
> > @ClaudiaComito I have added a check to confirm it is a scalar dndarray , Is anything wrong over there?
> 
> @Dhruv454000 no nothing wrong, we can simplify things a bit (see above). When you're done with that, we can run the GPU tests and check if the test coverage is sufficient.
> 
> Also could you please update the changelog? There's a similar entry already, you can add to that one.

Done.

**Comment by ClaudiaComito** [comment]
run tests

**Comment by ClaudiaComito** [comment]
run tests

**Comment by Dhruv454000** [comment]
Should we change that .format part?

**Comment by ClaudiaComito** [comment]
run tests

**Comment by ClaudiaComito** [comment]
> Should we change that .format part with fstrings

I have no strong opinion on that, I'm hoping for the GPU tests to run through so we can optimize the test coverage if necessary

**Comment by ClaudiaComito** [comment]
@Dhruv454000 alright, last thing now is expand the tests (`test_sanitation`) for the check you introduced. Here's the code that's currently [not covered by tests](https://codecov.io/gh/helmholtz-analytics/heat/compare/febe4fa3c0755936aa0dd10e64762497ac33b03a...994134e5a0ede9bc95b4c5d5ac1c8954c189235e/diff)

**Comment by Dhruv454000** [comment]
> @Dhruv454000 alright, last thing now is expand the tests (`test_sanitation`) for the check you introduced. Here's the code that's currently [not covered by tests](https://codecov.io/gh/helmholtz-analytics/heat/compare/febe4fa3c0755936aa0dd10e64762497ac33b03a...994134e5a0ede9bc95b4c5d5ac1c8954c189235e/diff)

Sure , got it.

**Comment by Dhruv454000** [comment]
@ClaudiaComito not sure what's happening , I have added test.

**Comment by Dhruv454000** [comment]
> @ClaudiaComito not sure what's happening , I have added test.

**Comment by Dhruv454000** [comment]
> @ClaudiaComito not sure what's happening , I have added test.

**Comment by Dhruv454000** [comment]
Any issue @ClaudiaComito

**Comment by Dhruv454000** [comment]
@ClaudiaComito srry for confusion done above , now its done.
Also should we add test and condition to check whether it is DNDarray or not

File: heat/core/sanitation.py [hunk 3]
```diff
@@ -343,6 +343,17 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
     x : DNDarray
         with `x.ndim = 0`
     """
-    return factories.array(
-        x.larray.unsqueeze(0), dtype=x.dtype, split=x.split, comm=x.comm, device=x.device
+    if x.ndim != 0:
```
Comment (ClaudiaComito) [inline heat/core/sanitation.py:346]:
```suggestion
    if x.ndim != 0:
       if x.ndim == 1 and x.gnumel() == 1:
          return x
```

File: heat/core/sanitation.py [hunk 4]
```diff
@@ -343,6 +343,17 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
...
+        raise ValueError(
+            "Input needs to be a scalar DNDarray,but was found to be {}d DNDarray".format(x.ndim)
+        )
+    a = x.larray.unsqueeze(0)
+    return DNDarray(
+        a,
```
Comment (ClaudiaComito) [inline heat/core/sanitation.py:352]:
```suggestion
    return DNDarray(
        x.larray.unsqueeze(0),
```

File: heat/core/tests/test_sanitation.py [hunk 5]
```diff
@@ -51,6 +51,9 @@ def test_sanitize_sequence(self):
 
     def test_scalar_to_1d(self):
         ht_scalar = ht.array(8)
+        wrong_dim = ht.array([8])
+        with self.assertRaises(ValueError):
```
Comment (ClaudiaComito) [inline heat/core/tests/test_sanitation.py:55]:
```suggestion
        wrong_dim = ht.array([8, 1])
        with self.assertRaises(ValueError):
```

File: heat/core/tests/test_sanitation.py [hunk 6]
```diff
@@ -51,6 +51,9 @@ def test_sanitize_sequence(self):
...
+            ht.scalar_to_1d(wrong_dim)
```
Comment (ClaudiaComito) [inline heat/core/tests/test_sanitation.py:56]:
```suggestion
            ht.scalar_to_1d(wrong_dim)
         ht_1d_scalar =  ht.array([8])
         
```

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Just a few changes needed, thanks @Dhruv454000 !

File: heat/core/sanitation.py [hunk 7]
```diff
@@ -343,6 +343,18 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
     x : DNDarray
         with `x.ndim = 0`
     """
-    return factories.array(
-        x.larray.unsqueeze(0), dtype=x.dtype, split=x.split, comm=x.comm, device=x.device
+    if x.ndim != 0:
+       if x.ndim == 1 and x.gnumel() == 1:
+          return x
+        raise ValueError(
```
Comment (ClaudiaComito) [inline heat/core/sanitation.py:349]:
```suggestion
       raise ValueError(
```

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]

File: heat/core/sanitation.py [hunk 8]
```diff
@@ -343,6 +343,18 @@ def scalar_to_1d(x: DNDarray) -> DNDarray:
     x : DNDarray
         with `x.ndim = 0`
     """
-    return factories.array(
-        x.larray.unsqueeze(0), dtype=x.dtype, split=x.split, comm=x.comm, device=x.device
+    if x.ndim != 0:
+        if x.ndim == 1 and x.gnumel() == 1:
```
Comment (ClaudiaComito) [inline heat/core/sanitation.py:347]:
```suggestion
        if x.ndim == 1 and x.gnumel == 1:
```

**Review by ClaudiaComito — APPROVED** [review]
Thanks a lot @Dhruv454000 and apologies it took us so long...

**Pull request merged**